<!-- MK -->

<launch>
  	<!-- declare arguments to be passed in -->
  	<arg name="qcID" /> 
  	<arg name="firmware"/>
	<arg name="rotortype"/>
	<arg name="rawAccOffsetX"/>
	<arg name="rawAccOffsetY"/>
	<arg name="landHeight"/>
	<arg name="initialMass"/>
	<arg name="takeOffHeight"/>

<!-- telekyb core  -->
	<node name="TeleKybCore_$(arg qcID)" pkg="telekyb_core" type="telekyb_core_node">
		<param name="TeleKybCore/tRobotID" value="$(arg qcID)" />

		<param name="StateEstimatorController/tPublishRosTransform" value="false"/>
		
		<param name="StateEstimatorController/tPluginLookupName" value="state_estimators_plugin::EurocGroundTruthStateEstimator"/>
		<param name="EurocGroundTruthStateEstimator/tEurocGroundTruthSeTopicName" value="/firefly_$(arg qcID)/ground_truth/pose" />

<!-- 				<param name="StateEstimatorController/tPluginLookupName" value="state_estimators_plugin::EurocStateEstimator"/> -->
<!-- 				<param name="EurocStateEstimator/tEurocStateEstimatorSeTopicName" value="/firefly_$(arg qcID)/filteredState" /> -->

		
		<param name="TrajectoryController/tPluginLookupName" value="trajectory_trackers_plugin::EurocTrajectoryTracker"/>
		<param name="EurocTrajectoryTracker/tCommandsTopic" value="/TeleKyb/$(arg qcID)/Commands" />
		<param name="EurocTrajectoryTracker/tTaskNumber" value="31" />
		
		
		<param name="PositionControl/tDerivGain" value="2.3" />
		<param name="PositionControl/tXDerivGain" value="1.0" />
		<param name="PositionControl/tYDerivGain" value="0.9" />
		<param name="PositionControl/tPropGain" value="3.5" />
		<param name="PositionControl/tXPropGain" value="0.65" />
		<param name="PositionControl/tYPropGain" value="0.8" />
		<param name="PositionControl/tIntegGain" value="2.2" />
		<param name="PositionControl/tXIntegGain" value="0.8" />
		<param name="PositionControl/tYIntegGain" value="0.8" />
		<param name="PositionControl/tSatIntTerm" value="1.0" />
		
		<param name="PositionControl/tZPropGain" value="25.5" />
		<param name="PositionControl/tZDerivGain" value="11.9" />
		<param name="PositionControl/tZIntegGain" value="9.0" />
		<param name="PositionControl/tGravity" value="9.8" />
		
		<param name="tBrakeInPosition" value="true" />

		<param name="PositionControl/tDerGainVelMode" value="4" />
		<param name="PositionControl/tIntegVelGain" value="2.0" />
		<param name="PositionControl/tSatIntVelTerm" value="4.0" />
		
		<param name="StandardMassEstim/tInitialMass" value="$(arg initialMass)"/>
		
		<!--<param name="tCenterValueX" value="512" />
		<param name="tCenterValueY" value="512" />-->
		<param name="tTakeOffDestination" value="[0.0, 0.0, $(arg takeOffHeight)]"/>
		<param name="tTakeOffVertically" value="false"/>
		<param name="tLandDestinationHeight" value="$(arg landHeight)" />
		<param name="tFlyToDestination" value="[1.0, 1.0, 0.0]"/>
		
		<param name="TrajectoryProcessorController/tTrajectoryModules" value='["tk_trajprocessor/PMObsAvoid"]'/>
		<param name="tVelocityInputEnabled" value="true"/>
		<param name="tObsPointsTopicName" value="/TeleKyb/Obstacle_$(arg qcID)/ObstaclePoints"/>
	</node>


	<!-- <node name="PoseKalmanFilter_$(arg qcID)" pkg="tk_modelga" type="poseKalmanFilter">
		<param name="tInitialZCoordinate" value="0.08"/>
		<remap from="/firefly/imu" to="/firefly_$(arg qcID)/imu"/>
		<remap from="/firefly/pose1" to="/firefly_$(arg qcID)/sensors/pose1"/>
		<remap from="/firefly/pose2" to="/firefly_$(arg qcID)/sensors/pose2"/>
		<remap from="/firefly/ubiasedImu" to="/firefly_$(arg qcID)/ubiasedImu"/>
		<remap from="/firefly/filteredState" to="/firefly_$(arg qcID)/filteredState"/>
	</node> -->


	<!-- Obstacle Provider -->
	<node name="Obstacle_$(arg qcID)" pkg="tk_obstacle" type="obstacle_provider" respawn="true">
		<param name="ObstacleProviderController/tObstacleProviders" value='[ "tk_obstacle/Octomap" ]'/>
		<param name="ObstacleProviderController/tTKStateTopicName" value="/TeleKyb/TeleKybCore_$(arg qcID)/Sensor/TKState" />
		<param name="Octomap/tFileName" value="$(find telekyb_gazebo)/resource/power_plant.bt" />
		<param name="Octomap/tNeighboroudRay" value="1.0" />
	</node>	
	
	<!-- Euroc Thing for the simulator -->
	<node name="euroc_solution_$(arg qcID)" pkg="tk_gazebo_interface" type="tk_gazebo_interface">
 	       <param name="robotName" value="firefly" />
 	       <param name="robotID" value="$(arg qcID)" />
		<remap from="/default/imu" to="/firefly_$(arg qcID)/imu"/>
		<remap from="/default/pose" to="/firefly_$(arg qcID)/pose1"/>
		<remap from="/TeleKyb/0/Commands" to="/TeleKyb/$(arg qcID)/Commands"/>
		<remap from="/default/command/attitude" to="/firefly_$(arg qcID)/command/attitude"/>
	</node>
	
</launch>
